### ISSUE TITLE
Steam window steals focus while opening

### ISSUE THREAD

**Issue by tapir** [ValveSoftware/steam-for-linux#3783]
1- User starts steam
2- Steam update phase kicks in
3- User does other stuff while waiting for update to finish
4- Steam switches to logging in phase and infuriates user by stealing the focus
5- User switches back to whatever he was doing
6- Steam window opens and steal backs the focus
7- User switched back to whatever he was doing
8- Steam ad window steals back the focus
9- User just wants to kill himself

**Comment by mojangsta** [comment]
Handling of focus is done by the window manager. Steam does not set any special focus hints/flags on its login windows if I'm not mistaken, only on the chat and bottom-right notifications per user preference.

If your window manager doesn't let you change how it handles focus of new windows, you might want to switch to one that does. I use i3-gaps-next with my own modifications for focus-handling and placement.

**Comment by alemansec** [comment]
I use openbox, and also have the steam client stealing focus on start, both "stable" and beta steam clients steals focus upon startup.

My window manager (openbox) is configured to not set focus to newly opened windows, but do mouse-follow, and it actually works quite well on well-designed/implemented applications.

steps to reproduce:
1. launch the steam client (i personnally start it from an xterm, so i can see the various stdout/stderr outputs).
2. switch to another x11 window (i personally move my mouse to another xterm/app/monitor) ; newly selected window get the focus
3. continue working in newly selected window, which has focus.
4. the steam client opens up, and ... STEALS FOCUS

expected behaviour:
my currently "selected"/used x11 window keepds the mouse focus

obtained behaviour:
the steam client just steals focus ...


No option within the steam client allows me to change this steam clietn behaviour.

**Comment by tapir** [comment]
Unfortunately with a backlog of 2K issues I'm very pessimistic about this issue ever getting fixed.

**Comment by PhoenixRiver** [comment]
The claim that the window manager can stop Steam is bogus... I am running Cinnamon (and used to run Xfce) and neither of these WM's could stop Steam stealing focus. My current window handling settings in Cinnamon (3.2.7) are:
Window Focus Mode: Mouse
Automatically raise focused windows: OFF
Bring windows which require attention to the current workspace: OFF
Prevent focus stealing: ON

Those are pretty explicit options to stop focus stealing in many contexts but Steam still manages it, THREE times! *rude*

My updated details (since the problem is still present in what is quite a different setup since I first submitted this problem - I've shifted distribution _and_ moved up a version on both the old one and the new one plus got a much newer kernel and been shifted to open source graphics drivers and changed window manager but this is still an issue!):

uname:
4.4.0-71-generic #92-Ubuntu SMP Fri Mar 24 12:59:01 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

Steam client application
Built: Mar 22 2017, at 23:45:04
Steam API: v017
Steam package version: 1490228413

Cinnnamon versions (latest for Linux Mint 18.1):
cinnamon 3.2.7+serena
cinnamon-common 3.2.7+serena
cinnamon-control-center 3.2.1+serena
cinnamon-control-center-data 3.2.1+serena
cinnamon-desktop-data 3.2.4+serena
cinnamon-I10n 3.2.2
cinnamon-screensaver 3.2.13+serena
cinnamon-session 3.2.0+serena
cinnamon-session-common 3.2.0+serena
cinnamon-settings-daemon 3.2.1+serena
cinnamon-themes 2016.12.07

P.S. I don't expect that this will be fixed (as with many other issues e.g. hugely dated libraries being used) and it's all the more reason to use better platforms like itch - it's just a pity that there are fewer professional dev's using others at the moment; though it will change in time as new dev's build themselves up on other more hospitable platforms.

**Comment by tapir** [comment]
It's been 3 years. Oh god please fix this issue and save our souls from rotting with rage.

**Comment by drudiger** [comment]
Lol, @kisak-valve keeps closing new issues over the years as duplicates and referencing this one, so they're aware it still exists, but they just won't spend the time to fix it.

Also, I initially didn't realize this was the Steam for Linux repo (found this issue on Google), and I actually have the issue on the Windows client. Doesn't look like they have a repo for that one though.

**Comment by robertc** [comment]
It's not just the main Steam window, the Friends window also steals focus repeatedly.  While we have the current issues with connecting to the Friends network it is extra annoying - after you click 'RETRY CONNECTION' and try to go back to what you were doing the window will steal focus a few seconds later, then it will steal focus again when it fails to connect.  This basically makes it very annoying to do anything else with your PC except click the button and wait.

**Comment by A-UNDERSCORE-D** [comment]
This continues to be an issue. _especially_ during updates, where the update window pops up, then once you click it, two zenity windows, followed by the loading window, followed by the library and news, followed by chat. All of which steal focus. It is _infuriating_ while trying to write code or type a message.

**Comment by Augustin82** [comment]
Valve, get it together and fix this.

**Comment by aseriesofdarkcaves** [comment]
I would very much appreciate a fix for this issue too.

**Comment by Cryotize** [comment]
Same

**Comment by TheBoctor** [comment]
Thought this was a bug in GNOME Shell or mutter when it happened during updates, which was not fun to try and troubleshoot. Steam is the only software on my system that steals focus like this.

Agree with the reasoning already stated here. Would appreciate a fix or a temporary DIY workaround; can give any information needed.

**Comment by tutacat** [comment]
Steam needs to forward to initial program-launch time, or the time of the interaction that caused a window to open
to GTK, and relevant window managing protocols. Instead, by not forwarding the time, the window manager just has to take the current time as it is, breaking focus-stealing prevention.
https://mail.gnome.org/archives/desktop-devel-list/2004-December/msg00306.html

**Comment by Zorrototo** [comment]
7 years later and Steam still STEALS FOCUS.

**Comment by drzetein** [comment]
fix this.

**Comment by AndrewDavis** [comment]
> Handling of focus is done by the window manager. Steam does not set any special focus hints/flags on its login windows if I'm not mistaken...

Okay, no, as a developer, I know this is BS. Whenever you go to create a window, there are always options relating to how a window presents itself. And, in fact, I remember one having to specifically go out of their way to make a window appear on top and steal focus. For instance, on Windows, there are API functions like [BringWindowToTop](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-bringwindowtotop) and [SetForegroundWindow](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setforegroundwindow).

So, in other words, Steam does this deliberately, with every single window. It's intentional. They haven't fixed it because they don't want to. We'll need to come up with our own fix/hack to prevent Steam from doing this.

**Comment by USN484259** [comment]
I have just made a **dirty hack** into the *Ubuntu Desktop Environment* to prevent **most** focus stealing issues (including Steam ones). However it involves modifying & recompiling system core libraries, so maybe too hard/dangerous for people lacking technical background. Have a try if you really **hate focus stealing** and could understand the hacking instructions.
https://github.com/USN484259/persistent-focus

**Comment by sedyh** [comment]
+1 for this. This is very annoying.

**Comment by CapricanDRJ** [comment]
Why is this still an issue? It shouldn't take this long to fix.

**Comment by junagu** [comment]
Oh, so not just me, then! Yes, agree with all of this - it's really, really annoying, no way to stop it happening and there are umpteen windows steals. I'm running XFCE on MX-Linux. Really hate the Steam client, it's horrible. Way too much crap in it. 

Please fix the windows focus stealing!

**Comment by kisak-valve** [comment]
Bug: Client doesn't provide correct last interaction timestamp when opening new windows.

Issue transferred from https://github.com/ValveSoftware/steam-for-linux/issues/9783.
@tutacat posted on 2023-07-03T12:41:29:

#### Your system information

* Steam client version (build number or date): 1688171965 (Sat, Jul 1 1:19 AM UTC), but affects versions at least back to 2015
* Distribution (e.g. Ubuntu): Fedora
* Opted into Steam client beta?: Yes
* Have you checked for system updates?: Yes
* Steam Logs: 
[steam-logs_1688171965_03072023.zip](https://github.com/ValveSoftware/steam-for-linux/files/11936761/steam-logs_1688171965_03072023.zip)
* GPU: AMD

Operating System Version:
- "Fedora release 38 (Thirty Eight)" (64 bit)
- Kernel Name: Linux
- Kernel Version: 6.0.5-200.fc36.x86_64
- X Server Vendor: The X.Org Foundation
- X Server Release: 12201009 (XWayland)
- X Window Manager: GNOME Shell
- Steam Runtime Version: steam-runtime_0.20230509.49499

Video Card:
- Driver: AMD AMD Radeon Graphics (renoir, LLVM 15.0.6, DRM 3.48, 6.0.5-200.fc36.x86_64)
- Driver Version: 4.6 (Compatibility Profile) Mesa 22.3.3
- OpenGL Version: 4.6
- Desktop Color Depth: 24 bits per pixel
- Monitor Refresh Rate: 59 Hz
- VendorID: 0x1002
- DeviceID: 0x1638
- Revision Not Detected
- Number of Monitors: 1
- Number of Logical Video Cards: 1

#### Please describe your issue in as much detail as possible:
Expected: When a window is opened, the program passes the causal interaction timestamp (i.e. launch, click), to the window manager.

Behaviour: Interaction timestamp is not passed, and the current timestamp is used.

#### Steps for reproducing this issue:

1. Launch Steam
2. Either before the splash window, or before the library window is shown, click on, or type in, another open window than Steam.
3. The new Steam window is focused, even if another interaction happened while Steam was loading.
- The window manager is not passed the correct window-open timestamp (i.e. Steam launching), so it doesn't know whether the new window should be focused or not.

**Comment by AndrewDavis** [comment]
Petition to change name from Steam to Steamroll because that's what it does to the other windows and focus

**Comment by Zorrototo** [comment]
This is VERY annoying issue, but to be fair if it is not fixed yet it is because this is very low priority due to the severeness of the issue.

**Comment by AndrewDavis** [comment]
Maybe they just didn't see the Issue or got distracted from it because they launched Steamroll and it steamrolled their screen.

**Comment by USN484259** [comment]
It seems that nowadays Steam client **steals focus** more frequently and aggressively. Sometimes the main window pops up for no reason when e.g. launching a game via taskbar icon. Today I've tried Steam client on a system without [my hack above](https://github.com/ValveSoftware/steam-for-linux/issues/3783#issuecomment-1220743983) appied, the **focus stealing** IS SO FVCKING CRAZY and I cannot help thinking that if Valve ever have the concept of *respect* in their mind.

There are so many applications that **steal focus** on different OS platforms. Some are unintentional, some should blame the GUI toolkit, also there're OS's design flaws. However, Steam client is obviously INTENTIONAL considering its timing and frequency of **focus stealing**

Maybe They just consider this as *work as expected* and won't fix, but cannot say that for whatever reason.

I think their arrogance come from the fact that Steam is one of the top (if not the only) game platforms on Linux. So no matter how Valve nauseates you, you still have to buy games on it.

AFAIK, to solve the problem, you have multiple choices:
1. Move to a game platform that works on Linux and won't **steal focus** which may not exist yet
2. Develop a game platform that works on Linux and won't **steal focus** if you have time and money
3. Move to other OS platform, voiding all your knowledge about Linux
4. Stop playing computer games
5. **Hack** the Steam program and being sued by Valve
6. **Hack** the OS at the risk of crashing the whole system
7. Fvck yourself and think that **focus stealing** is okay

Which will you choose ?

**Comment by Admonito** [comment]
![image](https://github.com/ValveSoftware/steam-for-linux/assets/116909372/c31717ee-2d88-4109-be4c-d41fd26feae7)
I have found a setting in Xfce4 which you could disable focus stealing 
Go to Applications>Settings>Tune Window Manager>Focus>Turn on prevent focus stealing
I have not tried this yet,but if it doesnt work i will delete this or Write under it that it doesnt work. :]

**Comment by Zorrototo** [comment]
Why do you even post when you admit you don't know what it does?

Spoiler alert, like other settings in other DE it will fail.

**Comment by Admonito** [comment]
It looks like after changing that setting made it not steal focus,but made the game unplayable for me due to lags and windows where still opening in the background (error windows and friend window)
I recommed to use Kde plasma if you have problems with steam stealing focus,when useing kde plasma i have never had a focus stealing issue. :]

**Comment by Zorrototo** [comment]
The issue exists in Plasma too, unless you increase the stealing focus prevention methods. But then you create other issue where windows don't get focus when they could/should.

What setting did you set your Plasma?

**Comment by Zorrototo** [comment]
Plasma had the issue.

**Comment by Admonito** [comment]
Weird. 🤔

**Comment by Admonito** [comment]
> Plasma had the issue.
> 
> //EDIT: since when are you using Plasma?

Around a month ago.

**Comment by awsms** [comment]
The absolute state of closed-source software

**Comment by tutacat** [comment]
> The absolute state of closed-source software

The best workaround is disable autostart or add -silent to the ~/.config/autostart/steam.desktop

**Comment by taernsietr** [comment]
> > The absolute state of closed-source software
> 
> The best workaround is disable autostart or add -silent to the ~/.config/autostart/steam.desktop

That isn't a workaround considering the issue is about the events that happen when Steam is *opened*, not just autostarts.

In fact, while we are discussing this under the steam-for-linux repo, this issue also happens on Windows, in much the same way - which is precisely why I don't understand why the devs won't even give us the actual reason they ignore the issue.

**Comment by awsms** [comment]
>this issue also happens on Windows, in much the same way

Mh I had no idea. Then the best workaround is to buy games on GOG.

**Comment by didli** [comment]
I consider this an issue too, but I found this especially annoying when steam is **closing**.
Steps to reproduce : 
1. play a game
2. end the game
3. quit steam
4. maximize your browser
5. steam steals focus to inform you (because that's obviously important to _know_) that it is - indeed - closing now.
6. user wants to kill steam.

**Comment by Bl4ckspell7** [comment]
Valve pls fix

**Comment by sull324** [comment]
Yep same issue,steam hugging the focus on kde plasma, have to do some witch voodoo magic switching tab process to restore my focus. Feel free to to use some of the money i spend on your platform to pay someone to fix that bug,or me actually if you cant do it.

**Comment by sinsinewave** [comment]
similar issue on multimonitor i3wm on Void Linux; steam not only steals focus on startup and completely randomly, but in the process also forces itself into the currently active workspace, regardless of monitor

**Comment by spoutn1k** [comment]
Same issue on both macos and gnome. How is this 8 years old ?

**Comment by hacheraw** [comment]
Same on Windows. Quite old unwanted behaviour but let's make more steam cards!

**Comment by mike-whitehurst** [comment]
Please review the priority of this. The impact may only be brief irritation but the frequency is every client launch. I think that should make it "fix for next release" priority.

**Comment by undeference** [comment]
Does it affect steam link? If not, expect a "fix" with the next major redesign.

**Comment by Legendary4226** [comment]
Yes pls fix this windows and linux, that's sooooo anoying you can't imagine

**Comment by marcokloeckler** [comment]
Hey there,
not sure if this fixes your problem, it fixed mine: Adding the '-silent' option for starting steam.
For example: `"C:\Program Files (x86)\Steam\steam.exe" -silent`

See addintional information at [the steam help page](https://help.steampowered.com/en/faqs/view/0188-6BB7-D467-08E1)

**Comment by Zorrototo** [comment]
What the! It works!

**Comment by hacheraw** [comment]
> Hey there, not sure if this fixes your problem, it fixed mine: Adding the '-silent' option for starting steam. For example: `"C:\Program Files (x86)\Steam\steam.exe" -silent`
> 
> See additional information at [the steam help page](https://help.steampowered.com/en/faqs/view/0188-6BB7-D467-08E1)

Yeah, it works, thanks. Maybe too silent for me now haha. I have to wait and click the client shortcut again (or traybar icon) for it to show up

**Comment by spacescientist** [comment]
> Hey there, not sure if this fixes your problem, it fixed mine: Adding the '-silent' option for starting steam. For example: `"C:\Program Files (x86)\Steam\steam.exe" -silent`
> 
> See additional information at [the steam help page](https://help.steampowered.com/en/faqs/view/0188-6BB7-D467-08E1)

Many thanks @marcokloeckler ; it's perfect in my case as well, it finally solves this.

I would have been nice if we had been informed earlier e.g. by @kisak-valve. I wonder how long this option has been there without us knowing...

**Comment by undeference** [comment]
> Hey there, not sure if this fixes your problem, it fixed mine: Adding the '-silent' option for starting steam. For example: `"C:\Program Files (x86)\Steam\steam.exe" -silent`
> 
> See additional information at [the steam help page](https://help.steampowered.com/en/faqs/view/0188-6BB7-D467-08E1)

This is not a fix but a workaround that works for many users. It has been posted, now please stop posting "this fixes it for me" or the like since it is spam for people interested in any (unlikely) progress.

**Comment by Managor** [comment]
Was this fixed on March 18th or is the issue still present?
https://store.steampowered.com/news/app/1675200/view/529841158837241541?l=english
> Fixed Steam Client window focusing multiple times on startup.

**Comment by echaskaris** [comment]
Nope, not for me
Ubuntu 22.04

**Comment by kozross** [comment]
Also for me. This is super annoying and I wish it would get fixed.

**Comment by modellking** [comment]
I also have similar symptoms on Gnome 49.9
